tush.c

```diff
@@ -255,357 +361,165 @@ int main(int argc, char** argv) {
-                            XDrawString(real_display, real_window, DefaultGC(real_display, real_s), ra.width / 20, 
-                                                            ra.height / 8, 
-                                                            str, len);
-                            break;
-                        }
-                    }
-                }
-            }
 
-            XFreeEventData(real_display, cookie);
-        } else {
+        /* 
+            if we check !XGetEventData here instead of waiting for the else, we can just skip the nesting
+
+            this makes the code a bit cleaner, not sure if it does much for the actual performance
+        */
+
+        if (!XGetEventData(real_display, cookie)) {
             if(event.type == xkbEventCode) {
                 XkbEvent *xkbEvent = (XkbEvent*)&event;
                 if (xkbEvent->any.xkb_type == XkbStateNotify) {
                     group = xkbEvent->state.group;
                 }
             }
+
+            continue;
         }
-    }
 
-    free(msg);
+        // Handle key press and release events
+        if (cookie->type == GenericEvent
+                && cookie->extension == xiOpcode
+                && (cookie->evtype == XI_RawKeyRelease /* not much reason to have this in it's own if statement */
+            || cookie->evtype == XI_RawKeyPress)) {
+                
+                XIRawEvent *ev = cookie->data;
+
+                KeySym s = XkbKeycodeToKeysym(
+                        real_display, ev->detail, group, 0 /*shift level*/);
+
+                if (NoSymbol == s) {
+                    if (group == 0) continue;
+                    else {
+                        s = XkbKeycodeToKeysym(real_display, ev->detail,
+                                0 /* base group */, 0 /*shift level*/);
+                        if (NoSymbol == s) continue;
+                    }
+                }    
+
+                change_opacity(real_display, &real_window, alpha);
+
+                /* 
+                    the switch is useless if there's only a default argument  
+                
+                    let's fix that!
+                */
+
+                /*
+                    keycodes should be used here because
+                    comparing the keycode is 100% faster
+                    than comparing strings
+                */
+
+                char* str;
+                size_t len = 1;
+
+                switch(s) {
+                    case XK_grave:
+                        str = "`"; break;
+                    case XK_comma:
+                        str = ","; break;
+                    case XK_slash:
+                        str = "/"; break;
+                    case XK_equal:
+                        str = "="; break;
+                    case XK_minus:
+                        str = "-"; break;
+                    case XK_space:
+                        str[0] = 'S'; break;
```

Comment:
`str = "S";` would be great.

Reply:
My mistake, I think it was "String" originally